src/Workspaces/Core/MSBuild/MSBuild/SerializationExtensions.cs

```diff
@@ -0,0 +1,67 @@
+            var analyzerReferenceChecksums = new AnalyzerReferenceChecksumCollection(projectInfo.AnalyzerReferences.SelectAsArray(ar => serializer.CreateChecksum(ar, default)).ToArray());
+
+            var documentChecksums = new DocumentChecksumCollection(projectInfo.Documents.SelectAsArray(di => di.GetCheckSum()).ToArray());
+            var additionalChecksums = new TextDocumentChecksumCollection(projectInfo.AdditionalDocuments.SelectAsArray(di => di.GetCheckSum()).ToArray());
+            var analyzerConfigDocumentChecksums = new AnalyzerConfigDocumentChecksumCollection(projectInfo.AnalyzerConfigDocuments.SelectAsArray(di => di.GetCheckSum()).ToArray());
+
+            return new ProjectStateChecksums(infoChecksum,
+                                             compilationChecksum,
+                                             parseOptionsChecksum,
+                                             documentChecksums,
+                                             projectReferenceChecksums,
+                                             metadataReferenceChecksums,
+                                             analyzerReferenceChecksums,
+                                             additionalChecksums,
+                                             analyzerConfigDocumentChecksums);
+        }
+
+        public static Checksum GetCheckSum(this DocumentInfo documentInfo)
+        {
+            return documentInfo.Attributes.GetCheckSum();
+        }
+
+        public static Checksum GetCheckSum<T>(this T @object)
+            where T : IObjectWritable
+        {
+            return @object switch
+            {
+                ProjectChecksumCollection x => Checksum.Create(WellKnownSynchronizationKind.Projects, x),
+                DocumentChecksumCollection x => Checksum.Create(WellKnownSynchronizationKind.Documents, x),
+                TextDocumentChecksumCollection x => Checksum.Create(WellKnownSynchronizationKind.TextDocuments, x),
+                AnalyzerConfigDocumentChecksumCollection x => Checksum.Create(WellKnownSynchronizationKind.AnalyzerConfigDocuments, x),
+                ProjectReferenceChecksumCollection x => Checksum.Create(WellKnownSynchronizationKind.ProjectReferences, x),
+                MetadataReferenceChecksumCollection x => Checksum.Create(WellKnownSynchronizationKind.MetadataReferences, x),
+                AnalyzerReferenceChecksumCollection x => Checksum.Create(WellKnownSynchronizationKind.AnalyzerReferences, x),
+                SolutionInfo.SolutionAttributes x => Checksum.Create(WellKnownSynchronizationKind.SolutionAttributes, x),
+                ProjectInfo.ProjectAttributes x => Checksum.Create(WellKnownSynchronizationKind.ProjectAttributes, x),
+                DocumentInfo.DocumentAttributes x => Checksum.Create(WellKnownSynchronizationKind.DocumentAttributes, x),
```

Comment:
Any reason this can't just call https://github.com/dotnet/roslyn/blob/dd21600f51cf5464b4721b69ab1c8e86bfa0c101/src/Workspaces/Core/Portable/Serialization/SerializationExtensions.cs#L16 or something similar?